### ISSUE TITLE
Support Header matching in Endpoint Routing

### ISSUE THREAD

**Issue by davidni** [dotnet/aspnetcore#21157]
### Is your feature request related to a problem? Please describe.
No. In the context of YARP, some routes must support header matching. If there is interest, I can submit a PR to contribute that here.

### Describe the solution you'd like
Our implementation handles the following:
* Match on the presence of a given header
* Match on a given value of a header (e.g. only match if header `x-example` has value `abc`, `def` or `ace`. Custom string comparison modes are supported)
* Match on a given prefix of a header value (e.g. only match if header `x-eample` starts with `abc`, `def` or `ace`. Custom string comparison modes are supported).

### Additional context
We have not implemented jump tables yet. We would hope the ASP .NET Core can take that on down the line if it seems relevant.

**Comment by davidni** [comment]
@rynowak fyi as you helped in our internal implementation.

**Comment by rynowak** [comment]
Yes, I think we'd be very interested in having your implementation contributed 👍

**Comment by anurse** [comment]
@davidni you were up to contribute this, right?

**Comment by davidni** [comment]
@anurse yep, I filed the issue to confirm there is still interest in it and avoid unnecessary work if not. Expect something by mid next week, though no guarantees due to DRI duties.

**Comment by tmbull** [comment]
I'm not sure if this is useful out side of reverse proxy applications, but I have a specialized reversed proxy that could use this feature.
